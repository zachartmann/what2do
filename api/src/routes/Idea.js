import { Router } from "express";
import { StatusCodes } from "http-status-codes";
import IdeaModel from "../models/ideaSchema";
import { isEmpty } from "lodash";
import mongoose from "mongoose";

const router = Router();

/**
 * GET: /ideas - retrieve idea(s)
 */

router.get("/ideas", async (req, res) => {
  let ideas;

  // /ideas?ids=id1,id2,id3 allows for a single query for multiple IDs
  if (!isEmpty(req.query) && !isEmpty(req.query.ids)) {
    const _ids = req.query.ids.split(",");
    const ids = _ids.map((id) => mongoose.Types.ObjectId(id));

    ideas = await IdeaModel.find({
      _id: {
        $in: ids,
      },
    }).exec();
  } else {
    ideas = await IdeaModel.find({}).exec();
  }

  return res.status(StatusCodes.OK).json(ideas);
});

/**
 * POST: /idea - create or update an idea
 */

router.post("/idea", async (req, res) => {
  if (!isEmpty(req.body)) {
    const {
      _id,
      content,
      upVotes,
      downVotes,
      upVoters,
      downVoters,
      pinned,
      user,
      commentIds,
    } = req.body;

    if (_id) {
      // Edit idea if record exists instead
      const updatedModel = {
        _id,
        content,
        upVotes,
        downVotes,
        upVoters,
        downVoters,
        pinned,
        user,
        commentIds,
        lastModified: new Date(),
      };

      try {
        await IdeaModel.findOneAndUpdate({ _id }, updatedModel).exec();

        return res.sendStatus(StatusCodes.OK);
      } catch (err) {
        return res.status(StatusCodes.INTERNAL_SERVER_ERROR).json(err);
      }
    }

    const ideaToCreate = new IdeaModel({
      content,
      upVotes,
      downVotes,
      upVoters,
      downVoters,
      pinned,
      user,
      commentIds,
      createdAt: new Date(),
    });

    try {
      await ideaToCreate.save();

      return res.status(StatusCodes.CREATED).json(ideaToCreate);
    } catch (err) {
      return res.status(StatusCodes.INTERNAL_SERVER_ERROR).json(err);
    }
  } else {
    return res.sendStatus(StatusCodes.BAD_REQUEST);
  }
});

/**
 * DELETE: /idea/:id - delete an idea
 */

router.delete("/idea/:id", async (req, res) => {
  const _id = req.params.id;

  try {
    await IdeaModel.deleteOne({ _id }).exec();

    return res.sendStatus(StatusCodes.OK);
  } catch (err) {
    return res.status(StatusCodes.NOT_FOUND).json(err);
  }
});

export default router;
