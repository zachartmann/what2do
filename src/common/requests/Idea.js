import axios from "axios";
import { ideasEndpoint, ideaEndpoint } from "./Endpoints";
import { postPollIdea } from "./Poll";

/**
 * Ideas
 */

export async function getIdeas(ids = null) {
  // If multiple Idea IDs are provided get all of them from DB
  if (ids) {
    try {
      const response = await axios.get(ideasEndpoint, {
        params: {
          ids: ids.toString(),
        },
      });
      return response;
    } catch (err) {
      throw err;
    }
  } else {
    try {
      const response = await axios.get(ideasEndpoint);
      return response;
    } catch (err) {
      throw err;
    }
  }
}

export async function updateIdea(_id, content, upVotes, downVotes, pinned) {
  try {
    const response = await axios.post(ideaEndpoint, {
      _id,
      content,
      upVotes,
      downVotes,
      pinned,
    });
    return response;
  } catch (err) {
    throw err;
  }
}

export async function deleteIdea(_id) {
  try {
    const response = await axios.delete(`${ideaEndpoint}/${_id}`);

    return response;
  } catch (err) {
    throw err;
  }
}

export async function postIdea(
  poll_id,
  content,
  upVotes,
  downVotes,
  pinned,
  user
) {
  try {
    // Create idea record in database
    user = user ? user : "Anonymous";
    const ideaResponse = await axios.post(ideaEndpoint, {
      content,
      upVotes,
      downVotes,
      upVoters: [],
      downVoters: [],
      pinned,
      user,
    });
    console.log(`Idea created by: ${user}`);

    // Link created idea to poll
    await postPollIdea(poll_id, ideaResponse.data._id);
    return ideaResponse;
  } catch (err) {
    throw err;
  }
}
