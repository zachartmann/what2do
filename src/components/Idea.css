.idea {
  border: 1px solid black;
  border-radius: 10px;
  margin-left: 0;
  margin-right: 0;
}

.idea-small-details {
  margin-top: -40px;
}