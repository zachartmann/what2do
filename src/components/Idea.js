import React, { useState } from "react";
import { formatDistanceStrict } from "date-fns";

import CommentBox from "./CommentBox";

const Idea = ({ idea }) => {
  // Downvoted
  // -  Stroke red
  // -  Fill orangered? otherwise none

  // Upvoted
  // -  Stroke green
  // -  Fill lime otherwise none

  // Comment
  // -  Stroke .blue-icon
  // -  Fill white if not showing comments, otherwise lightskyblue

  // if (loggedInName in upVoters) {
  //   make it lime filled
  // }

  const [hidden, setHidden] = useState(true);
  const [commentCount, setCommentCount] = useState(0);
  const {
    content,
    upVotes,
    downVotes,
    upVoters,
    downVoters,
    pinned,
    user,
    lastModified,
    createdAt,
  } = idea;
  let metaLabel;

  if (lastModified) {
    const lastModifiedDate = new Date(lastModified);
    metaLabel = `Edited ${formatDistanceStrict(
      lastModifiedDate,
      new Date()
    )} ago`;
  } else if (createdAt) {
    const createdAtDate = new Date(createdAt);
    metaLabel = `Created ${formatDistanceStrict(
      createdAtDate,
      new Date()
    )} ago`;
  } else {
    metaLabel = "Unknown time created";
  }

  const incrementCommentCount = () => {
    setCommentCount(commentCount + 1);
  };

  const decrementCommentCount = () => {
    setCommentCount(commentCount - 1);
  };

  const toggleComments = () => {
    setHidden(!hidden);
  };

  const commentFill = hidden ? "white" : "lightskyblue";

  return (
    <div className="content">
      <div className="content-container idea">
        <div
          className="content-container flex-container"
          style={{ paddingBottom: "0px" }}
        >
          <div className="flex-component flex-70">
            <h3>{content}</h3>
          </div>
          <div className="flex-component flex-30 flex-end flex-container">
            <div className="flex-component">
              <svg
                className="h-6 w-6 icon"
                xmlns="http://www.w3.org/2000/svg"
                fill="orangered"
                viewBox="0 0 24 24"
                stroke="red"
              >
                <path
                  strokeLinecap="round"
                  strokeLinejoin="round"
                  strokeWidth="2"
                  d="M10 14H5.236a2 2 0 01-1.789-2.894l3.5-7A2 2 0 018.736 3h4.018a2 2 0 01.485.06l3.76.94m-7 10v5a2 2 0 002 2h.096c.5 0 .905-.405.905-.904 0-.715.211-1.413.608-2.008L17 13V4m-7 10h2m5-10h2a2 2 0 012 2v6a2 2 0 01-2 2h-2.5"
                />
              </svg>
            </div>
            <div className="flex-component">
              <p className="big centered">{upVotes - downVotes}</p>
            </div>
            <div className="flex-component">
              <svg
                className="h-6 w-6 icon"
                xmlns="http://www.w3.org/2000/svg"
                fill="lime"
                viewBox="0 0 24 24"
                stroke="green"
              >
                <path
                  strokeLinecap="round"
                  strokeLinejoin="round"
                  strokeWidth="2"
                  d="M14 10h4.764a2 2 0 011.789 2.894l-3.5 7A2 2 0 0115.263 21h-4.017c-.163 0-.326-.02-.485-.06L7 20m7-10V5a2 2 0 00-2-2h-.095c-.5 0-.905.405-.905.905 0 .714-.211 1.412-.608 2.006L7 11v9m7-10h-2M7 20H5a2 2 0 01-2-2v-6a2 2 0 012-2h2.5"
                />
              </svg>
            </div>
          </div>
        </div>
        <div className="content-container flex-container">
          <div className="flex-component flex-70 flex-between">
            <p className="small top idea-small-details">{user || "Anon"}</p>
            <p className="x-small top idea-small-details idea-meta-label">
              {metaLabel}
            </p>
          </div>
          <div className="flex-component flex-70 flex-end">
            {pinned ? (
              <svg
                className="h-6 w-6 icon blue-icon"
                xmlns="http://www.w3.org/2000/svg"
                fill="lightskyblue"
                viewBox="0 0 24 24"
              >
                <path
                  fillRule="evenodd"
                  d="M4.293 15.707a1 1 0 010-1.414l5-5a1 1 0 011.414 0l5 5a1 1 0 01-1.414 1.414L10 11.414l-4.293 4.293a1 1 0 01-1.414 0zm0-6a1 1 0 010-1.414l5-5a1 1 0 011.414 0l5 5a1 1 0 01-1.414 1.414L10 5.414 5.707 9.707a1 1 0 01-1.414 0z"
                  clipRule="evenodd"
                />
              </svg>
            ) : null}
            <svg
              onClick={toggleComments}
              className="h-6 w-6 icon blue-icon comment-icon button-icon"
              xmlns="http://www.w3.org/2000/svg"
              fill={commentFill}
              viewBox="0 0 24 24"
            >
              <path
                strokeLinecap="round"
                strokeLinejoin="round"
                strokeWidth="2"
                d="M7 8h10M7 12h4m1 8l-4-4H5a2 2 0 01-2-2V6a2 2 0 012-2h14a2 2 0 012 2v8a2 2 0 01-2 2h-3l-4 4z"
              />
            </svg>
            <p className="top" style={{ paddingLeft: "7px" }}>
              {commentCount}
            </p>
          </div>
        </div>
        <div className="content-container">
          <CommentBox
            hidden={hidden}
            commentCount={commentCount}
            increment={incrementCommentCount}
            decrement={decrementCommentCount}
          />
        </div>
      </div>
    </div>
  );
};

export default Idea;
