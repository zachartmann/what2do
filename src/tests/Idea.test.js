import { shallow } from "enzyme";
import Idea from "../components/Idea";

const dummyIdea = {
  _id: "507f191e810c19729de860ea",
  content: "content",
  upVotes: 2,
  downVotes: 3,
  upVoters: [
    {
      _id: "507f191e810c19729de860eb",
      name: "Kevin",
      password: "Password",
    },
  ],
  downVoters: [
    {
      _id: "507f191e810c19729de860ec",
      name: "Kevin",
      password: "Password",
    },
  ],
  pinned: true,
  user: "Kevin",
};

describe("Idea component", () => {
  it("should render correctly with given idea", () => {
    const component = shallow(<Idea key={dummyIdea._id} idea={dummyIdea} />);

    expect(component).toMatchSnapshot();
  });
});
