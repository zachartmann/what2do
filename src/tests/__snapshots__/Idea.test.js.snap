// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Idea component should render correctly with given idea 1`] = `
<div
  className="content"
>
  <div
    className="content-container idea"
    onMouseEnter={[Function]}
    onMouseLeave={[Function]}
  >
    <div
      onClick={[Function]}
      style={
        Object {
          "height": 0,
        }
      }
    >
      <svg
        className="icon blue-icon pin-icon"
        fill="yellow"
        stroke="currentColor"
        viewBox="0 0 24 24"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M11.049 2.927c.3-.921 1.603-.921 1.902 0l1.519 4.674a1 1 0 00.95.69h4.915c.969 0 1.371 1.24.588 1.81l-3.976 2.888a1 1 0 00-.363 1.118l1.518 4.674c.3.922-.755 1.688-1.538 1.118l-3.976-2.888a1 1 0 00-1.176 0l-3.976 2.888c-.783.57-1.838-.197-1.538-1.118l1.518-4.674a1 1 0 00-.363-1.118l-3.976-2.888c-.784-.57-.38-1.81.588-1.81h4.914a1 1 0 00.951-.69l1.519-4.674z"
          strokeLinecap="round"
          strokeLinejoin="round"
          strokeWidth="2"
        />
      </svg>
    </div>
    <div
      className="content-container flex-container"
      style={
        Object {
          "paddingBottom": "0px",
        }
      }
    >
      <div
        className="flex-component flex-70"
      >
        <h3>
          content
        </h3>
      </div>
      <VotingMechanism
        idea={
          Object {
            "_id": "507f191e810c19729de860ea",
            "commentIds": Array [
              "0",
            ],
            "content": "content",
            "downVoters": Array [
              Object {
                "_id": "507f191e810c19729de860ec",
                "name": "Kevin",
                "password": "Password",
              },
            ],
            "downVotes": 3,
            "pinned": true,
            "upVoters": Array [
              Object {
                "_id": "507f191e810c19729de860eb",
                "name": "Kevin",
                "password": "Password",
              },
            ],
            "upVotes": 2,
            "user": "Kevin",
          }
        }
        user={
          Object {
            "_id": "615689f3c7568fc6aeff9000",
            "name": null,
            "password": "",
          }
        }
      />
    </div>
    <div
      className="content-container flex-container"
    >
      <div
        className="flex-component flex-70 flex-between"
      >
        <p
          className="small top idea-small-details"
        >
          Kevin
        </p>
        <p
          className="x-small top idea-small-details idea-meta-label"
        >
          Unknown time created
        </p>
      </div>
      <div
        className="flex-component flex-70 flex-end"
      >
        <svg
          className="h-6 w-6 icon blue-icon comment-icon button-icon"
          fill="white"
          onClick={[Function]}
          viewBox="0 0 24 24"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M7 8h10M7 12h4m1 8l-4-4H5a2 2 0 01-2-2V6a2 2 0 012-2h14a2 2 0 012 2v8a2 2 0 01-2 2h-3l-4 4z"
            strokeLinecap="round"
            strokeLinejoin="round"
            strokeWidth="2"
          />
        </svg>
        <p
          className="top"
          style={
            Object {
              "paddingLeft": "7px",
            }
          }
        >
          1
        </p>
      </div>
    </div>
    <div
      className="content-container"
    >
      <CommentBox
        decrement={[Function]}
        hidden={true}
        increment={[Function]}
        initialCommentIds={
          Array [
            "0",
          ]
        }
        updateIdeaCommentIds={[Function]}
      />
      <Modal
        idea={
          Object {
            "_id": "507f191e810c19729de860ea",
            "commentIds": Array [
              "0",
            ],
            "content": "content",
            "downVoters": Array [
              Object {
                "_id": "507f191e810c19729de860ec",
                "name": "Kevin",
                "password": "Password",
              },
            ],
            "downVotes": 3,
            "pinned": true,
            "upVoters": Array [
              Object {
                "_id": "507f191e810c19729de860eb",
                "name": "Kevin",
                "password": "Password",
              },
            ],
            "upVotes": 2,
            "user": "Kevin",
          }
        }
        onClose={[Function]}
        onSubmit={[Function]}
        show={false}
        title="Edit Your Idea?"
      >
        <textarea
          className="comment-textarea"
          cols="50"
          maxLength="145"
          onChange={[Function]}
          placeholder="Enter your new idea"
          rows="3"
          value=""
        />
      </Modal>
    </div>
  </div>
</div>
`;
